File: strings/316.RemoveDuplicateLetters.go
package main

import "fmt"

func removeDuplicateLetters(s string) string {
	count := [26]int{}
	used := [26]bool{}
	res := []rune{}
	for i := range s {
		count[s[i]-'a']++
	}

	for i := range s {
		count[s[i]-'a']--
		if used[s[i]-'a'] {
			continue
		}
		for len(res) > 0 && res[len(res)-1] > rune(s[i]) && count[res[len(res)-1]-'a'] > 0 {
			used[res[len(res)-1]-'a'] = false
			res = res[:len(res)-1]
		}
		res = append(res, rune(s[i]))
		used[s[i]-'a'] = true
	}
	return string(res)
}

func main() {

	// Input: "bcabc"
	// Output: "abc"

	// Input: "cbacdcbc"
	// Output: "acdb"

	// Input: "cbbbcaa"
	// Output: "bca"

	// Input: "bbcaac""
	// Output: "bac"
	s := "bcabc"
	fmt.Println(removeDuplicateLetters(s))

	s = "cbacdcbc"
	fmt.Println(removeDuplicateLetters(s))

	s = "cbbbcaa"
	fmt.Println(removeDuplicateLetters(s))

	s = "bbcaac"
	fmt.Println(removeDuplicateLetters(s))
}

File: strings/316.RemoveDuplicateLetters.py
from typing import List


class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        # c = Counter(s)
        c = [0] * 26
        use = [0] * 26
        res: List[str] = []
        for i in s:
            c[ord(i) - ord("a")] += 1

        for i in s:
            c[ord(i) - ord("a")] -= 1

            if use[ord(i) - ord("a")]:
                continue

            while len(res) > 0 and res[-1] > i and c[ord(res[-1]) - ord("a")] > 0:
                use[ord(res[-1]) - ord("a")] = False
                res.remove(res[-1])
            res.append(i)
            use[ord(i) - ord("a")] = True

        return "".join(res)


if __name__ == "__main__":
    # Input: "bcabc"
    # Output: "abc"

    # Input: "cbacdcbc"
    # Output: "acdb"
    s = Solution()
    print(s.removeDuplicateLetters("bcabc"))
    print(s.removeDuplicateLetters("cbacdcbc"))

    print(s.removeDuplicateLetters("cbbbcaa"))  # "bca"
